fish-rust/src/builtins/math.rs

```diff
@@ -0,0 +1,319 @@
+        print_help: false,
+        scale: DEFAULT_SCALE,
+        base: 10,
+    };
+
+    let mut have_scale = false;
+
+    let mut w = wgetopter_t::new(SHORT_OPTS, LONG_OPTS, args);
+    while let Some(c) = w.wgetopt_long() {
+        match c {
+            's' => {
+                let optarg = w.woptarg.unwrap();
+                have_scale = true;
+                // "max" is the special value that tells us to pick the maximum scale.
+                opts.scale = if optarg == "max"L {
+                    15
+                } else if let Ok(base) = fish_wcstoi(optarg) {
+                    base
+                } else {
+                    streams.err.append(wgettext_fmt!(
+                        "%ls: %ls: invalid base value\n",
+                        cmd,
+                        optarg
+                    ));
+                    return Err(STATUS_INVALID_ARGS);
+                };
+            }
+            'b' => {
+                let optarg = w.woptarg.unwrap();
+                opts.base = if optarg == "hex"L {
+                    16
+                } else if optarg == "octal"L {
+                    8
+                } else if let Ok(base) = fish_wcstoi(optarg) {
+                    base
+                } else {
+                    streams.err.append(wgettext_fmt!(
+                        "%ls: %ls: invalid base value\n",
+                        cmd,
+                        optarg
+                    ));
+                    return Err(STATUS_INVALID_ARGS);
+                };
+            }
+            'h' => {
+                opts.print_help = true;
+            }
+            ':' => {
+                builtin_missing_argument(parser, streams, cmd, args[w.woptind - 1], print_hints);
+                return Err(STATUS_INVALID_ARGS);
+            }
+            '?' => {
+                // For most commands this is an error. We ignore it because a math expression
+                // can begin with a minus sign.
+                return Ok((opts, w.woptind - 1));
+            }
+            _ => {
+                panic!("unexpected retval from wgeopter.next()");
+            }
+        }
+    }
+
+    if have_scale && opts.scale != 0 && opts.base != 10 {
+        streams.err.append(wgettext_fmt!(
+            BUILTIN_ERR_COMBO2,
+            cmd,
+            "non-zero scale value only valid
+            for base 10"
+        ));
+        return Err(STATUS_INVALID_ARGS);
+    }
+
+    Ok((opts, w.woptind))
+}
+
+/// We read from stdin if we are the second or later process in a pipeline.
+fn use_args_from_stdin(streams: &io_streams_t) -> bool {
+    streams.stdin_is_directly_redirected()
+}
+
+/// Get the arguments from stdin.
+fn get_arg_from_stdin(streams: &io_streams_t) -> Option<WString> {
+    let mut s = WString::new();
+    loop {
+        let mut buf = [0];
+        let c = match read_blocked(streams.stdin_fd().unwrap().as_raw_fd(), &mut buf) {
+            1 => char::from(buf[0]),
```

Comment:
You are converting each input byte to a wide character and appending it to the string.
This may break on multibyte characters.
To be safe we should read into a `Vec<u8>` and convert it with `str2wcstring`, as the C++ version did.